src/tile.cpp

```diff
@@ -464,6 +470,26 @@ void Tile::onRemoveTileItem(const SpectatorHashSet& spectators, const std::vecto
 	for (Creature* spectator : spectators) {
 		spectator->onRemoveTileItem(this, cylinderMapPos, iType, item);
 	}
+
+	if (!hasFlag(TILESTATE_PROTECTIONZONE)) {
+		auto it = getItemList();
+		if (it->size() == 0) { // it->empty() does not work for some strange reason here
+			g_game.removeCleanTile(this);
+			return;
+		}
+
+		bool ret = false;
+		for (auto toCheck : *it) {
+			if (toCheck->isCleanable()) {
+				ret = true;
+				break;
+			}
+		}
+
+		if (!ret) {
+			g_game.removeCleanTile(this);
```

Comment:
Instead of doing this, you can short circuit inside `for` directly:

```cpp
for (auto toCheck : *it) {
	if (toCheck->isCleanable()) {
		g_game.removeCleanTile(this);
		break;
	}
}
```

Reply:
this does logic wise not work, only if all items in the stack of the tile return false, then it should remove the tile, else it should still clean it.